File: python/cortex_sdk/session.py
from __future__ import annotations

import json
import uuid
from typing import Callable

from .constants import SCHEMA_VERSION
from .errors import make_error, lookup_error, CortexError
from .transport import Transport
from .types import (
    CortexMessage,
    EscalationReplyAction,
    EscalationReplyContent,
    RuntimeBootstrap,
    SessionState,
)


_TERMINAL_STATES: set[SessionState] = {
    "COMPLETED",
    "FAILED",
    "STOPPED",
    "TIMEOUT",
    "CANCELLED",
}
_LIFECYCLE_STATE_MAP: dict[str, SessionState] = {
    "active": "ACTIVE",
    "waiting": "WAITING",
    "completed": "COMPLETED",
    "failed": "FAILED",
    "stopped": "STOPPED",
    "timeout": "TIMEOUT",
    "cancelled": "CANCELLED",
}


class SessionManager:
    """Manages session state and envelope construction/parsing."""

    def __init__(
        self,
        on_message: Callable[[CortexMessage], None],
        on_fatal_error: Callable[[CortexError], None],
    ) -> None:
        self._on_message = on_message
        self._on_fatal_error = on_fatal_error

        self._session_id: str | None = None
        self._session_state: SessionState = "CREATED"
        self._last_seq: int = 0
        self._transport: Transport | None = None
        self._send_timeout: float = 10.0

    def _set_session_state(self, next_state: SessionState) -> None:
        if self._session_state in _TERMINAL_STATES and self._session_state != next_state:
            return
        self._session_state = next_state

    # ── accessors ─────────────────────────────────────────────────────────────

    @property
    def session_id(self) -> str | None:
        return self._session_id

    @property
    def session_state(self) -> SessionState:
        return self._session_state

    @property
    def last_seq(self) -> int:
        return self._last_seq

    # ── configuration ─────────────────────────────────────────────────────────

    def set_transport(self, transport: Transport, send_timeout: float) -> None:
        self._transport = transport
        self._send_timeout = send_timeout

    # ── outbound helpers ──────────────────────────────────────────────────────

    def _build_envelope(
        self, type_: str, payload: dict[str, object]
    ) -> dict[str, object]:
        env: dict[str, object] = {
            "type": type_,
            "schema": SCHEMA_VERSION,
            "payload": payload,
            "ts": _now_iso(),
        }
        if self._session_id:
            env["session_id"] = self._session_id
        return env

    async def _send(self, envelope: dict[str, object]) -> None:
        if self._transport is None:
            raise make_error("transport_send_timeout", "No transport configured")
        await self._transport.send(envelope)

    async def send_init(self, bootstrap: RuntimeBootstrap) -> None:
        """Send system::init — intentionally omits session_id."""
        self._session_state = "INITIALIZING"
        envelope: dict[str, object] = {
            "type": "system::init",
            "schema": SCHEMA_VERSION,
            "payload": dict(bootstrap),
            "meta": {"client_msg_id": f"cli_init_{uuid.uuid4()}"},
            "ts": _now_iso(),
        }
        await self._send(envelope)

    async def send_resync(self) -> None:
        await self._send(
            self._build_envelope("system::resync", {"last_seq": self._last_seq})
        )

    async def send_stop(self) -> None:
        await self._send(self._build_envelope("sandbox::stop", {}))

    async def send_chat_message(
        self, content: str, attachments: list[str] | None
    ) -> None:
        payload: dict[str, object] = {"content": content, "role": "user"}
        if attachments:
            payload["meta"] = {"attachments": attachments}
        await self._send(self._build_envelope("chat::message", payload))

    async def send_escalation_reply(
        self,
        escalation_id: str,
        wait_token: str,
        action: EscalationReplyAction,
        content: EscalationReplyContent | None = None,
        meta: dict[str, object] | None = None,
    ) -> None:
        payload = _build_escalation_reply_payload(
            escalation_id=escalation_id,
            wait_token=wait_token,
            action=action,
            content=content,
            meta=meta,
        )
        await self._send(self._build_envelope("escalation::reply", payload))

    # ── inbound handling ──────────────────────────────────────────────────────

    def handle_incoming(self, raw: str) -> None:
        """Parse a raw JSON frame and dispatch callbacks."""
        try:
            parsed: dict[str, object] = json.loads(raw)
            msg: CortexMessage = parsed  # type: ignore[assignment]
        except Exception:
            return  # malformed frame — ignore

        # Learn session_id from first server message
        if not self._session_id and msg.get("session_id"):
            self._session_id = msg["session_id"]
            if self._session_state not in _TERMINAL_STATES:
                self._session_state = "ACTIVE"

        # Track highest sequence number
        seq = msg.get("seq")
        if isinstance(seq, int) and seq > self._last_seq:
            self._last_seq = seq

        self._update_session_state(msg)

        # Fatal error detection
        if msg.get("type") == "system::error":
            self._handle_system_error(msg["payload"])

        self._on_message(msg)

    def _handle_system_error(self, payload: dict[str, object]) -> None:
        code = payload.get("code", "session_terminal")
        message = payload.get("message", "Runtime error")
        entry = lookup_error(str(code))
        if entry and entry.fatal:
            self._set_session_state("FAILED")
            err = make_error(str(code), str(message))
            self._on_fatal_error(err)

    def _update_session_state(self, msg: CortexMessage) -> None:
        msg_type = msg.get("type")

        if msg_type == "sandbox::snapshot":
            state = msg["payload"].get("state")
            if isinstance(state, str) and state.lower() == "waiting":
                self._set_session_state("WAITING")
            return

        if msg_type == "sandbox::lifecycle":
            status = msg["payload"].get("status")
            if isinstance(status, str):
                next_state = _LIFECYCLE_STATE_MAP.get(status.lower())
                if next_state is not None:
                    self._set_session_state(next_state)
            return

        if msg_type == "system::error":
            code = msg["payload"].get("code")
            if isinstance(code, str):
                entry = lookup_error(code)
                if entry and entry.fatal:
                    self._set_session_state("FAILED")


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def _is_plain_object(value: object | None) -> bool:
    return isinstance(value, dict)


def _validate_escalation_reply(
    escalation_id: str,
    wait_token: str,
    action: str,
    content: EscalationReplyContent | None,
) -> None:
    if not isinstance(escalation_id, str) or escalation_id.strip() == "":
        raise make_error("transport_protocol_violation", "escalation_id is required")
    if not isinstance(wait_token, str) or wait_token.strip() == "":
        raise make_error("transport_protocol_violation", "wait_token is required")
    if action not in {"continue", "operator_input", "reply_user"}:
        raise make_error("transport_protocol_violation", f"Unsupported escalation reply action: {action}")
    if action == "continue" and content is None:
        return
    if isinstance(content, str):
        return
    if _is_plain_object(content):
        return
    raise make_error(
        "transport_protocol_violation",
        f"content must be a string or object for escalation action {action}",
    )


def _build_escalation_reply_payload(
    *,
    escalation_id: str,
    wait_token: str,
    action: EscalationReplyAction,
    content: EscalationReplyContent | None,
    meta: dict[str, object] | None,
) -> dict[str, object]:
    _validate_escalation_reply(escalation_id, wait_token, action, content)
    payload: dict[str, object] = {
        "escalation_id": escalation_id.strip(),
        "action": action,
        "wait_token": wait_token.strip(),
    }
    if content is not None:
        payload["content"] = content
    if meta is not None:
        payload["meta"] = meta
    return payload

File: js/src/session.ts
import { SCHEMA_VERSION } from './constants.js';
import { makeError, lookupError } from './errors.js';
import type { Transport } from './transport.js';
import type {
  CortexMessage,
  EscalationReplyAction,
  EscalationReplyContent,
  RuntimeBootstrap,
  SessionState,
} from './types.js';

const TERMINAL_STATES = new Set<SessionState>([
  'COMPLETED',
  'FAILED',
  'STOPPED',
  'TIMEOUT',
  'CANCELLED',
]);

const LIFECYCLE_STATE_MAP: Record<string, SessionState> = {
  active: 'ACTIVE',
  waiting: 'WAITING',
  completed: 'COMPLETED',
  failed: 'FAILED',
  stopped: 'STOPPED',
  timeout: 'TIMEOUT',
  cancelled: 'CANCELLED',
};

export interface SessionCallbacks {
  onMessage(msg: CortexMessage): void;
  onFatalError(err: Error): void;
}

export interface SessionController {
  sendInit(bootstrap: RuntimeBootstrap): Promise<void>;
  sendResync(): Promise<void>;
  sendStop(): Promise<void>;
  sendChatMessage(content: unknown, attachments: unknown[] | undefined, meta?: Record<string, unknown>): Promise<void>;
  sendEscalationReply(
    escalationId: string,
    waitToken: string,
    action: EscalationReplyAction,
    content: EscalationReplyContent | undefined,
    meta: Record<string, unknown> | undefined,
  ): Promise<void>;
  sendSystemTrigger(content: unknown, attachments: unknown[] | undefined): Promise<void>;
  sendTrigger(payload: Record<string, unknown>): Promise<void>;
  handleIncoming(data: string): void;
  get sessionId(): string | null;
  get sessionState(): SessionState;
  get lastSeq(): number;
  setTenantId(tenantId: string | null | undefined): void;
  setTransport(transport: Transport, sendTimeoutMs: number): void;
  reset(): void;
}

export function createSession(
  callbacks: SessionCallbacks,
): SessionController {
  let _sessionId: string | null = null;
  let _sessionState: SessionState = 'CREATED';
  let _lastSeq = 0;
  let _transport: Transport | null = null;
  let _sendTimeoutMs = 10000;
  let _tenantId: string | null = null;
  let _opened = false;

  function setSessionState(next: SessionState): void {
    if (TERMINAL_STATES.has(_sessionState) && _sessionState !== next) {
      return;
    }
    _sessionState = next;
  }

  function updateSessionStateFromMessage(msg: CortexMessage): void {
    if (msg.type === 'sandbox::snapshot') {
      const state = typeof msg.payload['state'] === 'string' ? msg.payload['state'].toLowerCase() : null;
      if (state === 'waiting') {
        setSessionState('WAITING');
      }
      return;
    }

    if (msg.type === 'sandbox::lifecycle') {
      const status = typeof msg.payload['status'] === 'string' ? msg.payload['status'].toLowerCase() : null;
      if (status && status in LIFECYCLE_STATE_MAP) {
        setSessionState(LIFECYCLE_STATE_MAP[status]);
      }
      return;
    }

    if (msg.type === 'system::error') {
      const code = typeof msg.payload['code'] === 'string' ? msg.payload['code'] : 'session_terminal';
      if (lookupError(code)?.fatal) {
        setSessionState('FAILED');
      }
    }
  }

  function send(envelope: Record<string, unknown>): Promise<void> {
    if (!_transport) return Promise.reject(new Error('No transport'));
    return _transport.send(envelope, _sendTimeoutMs);
  }

  function reset(): void {
    _sessionId = null;
    _sessionState = 'CREATED';
    _lastSeq = 0;
    _tenantId = null;
    _opened = false;
  }

  function buildEnvelope(type: string, payload: Record<string, unknown>): Record<string, unknown> {
    const env: Record<string, unknown> = {
      type,
      schema: SCHEMA_VERSION,
      payload,
      ts: new Date().toISOString(),
    };
    if (_sessionId) env['session_id'] = _sessionId;
    return env;
  }

  function makeClientMsgId(prefix: string): string {
    const cryptoRef = globalThis.crypto;
    if (cryptoRef && typeof cryptoRef.randomUUID === 'function') {
      return `${prefix}_${cryptoRef.randomUUID()}`;
    }
    return `${prefix}_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 12)}`;
  }

  function buildMessagePayload(
    content: unknown,
    attachments: unknown[] | undefined,
    meta?: Record<string, unknown>,
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = { content, role: 'user' };
    const combinedMeta: Record<string, unknown> = {};
    // meta merged first; official attachments param always wins
    if (meta) Object.assign(combinedMeta, meta);
    if (attachments && attachments.length > 0) combinedMeta['attachments'] = attachments;
    if (Object.keys(combinedMeta).length > 0) payload['meta'] = combinedMeta;
    return payload;
  }

  function isPlainObject(value: unknown): value is Record<string, unknown> {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
  }

  function validateEscalationReply(
    escalationId: string,
    waitToken: string,
    action: string,
    content: EscalationReplyContent | undefined,
  ): void {
    if (typeof escalationId !== 'string' || escalationId.trim() === '') {
      throw makeError('transport_protocol_violation', 'escalationId is required');
    }
    if (typeof waitToken !== 'string' || waitToken.trim() === '') {
      throw makeError('transport_protocol_violation', 'waitToken is required');
    }
    if (action !== 'continue' && action !== 'operator_input' && action !== 'reply_user') {
      throw makeError('transport_protocol_violation', `Unsupported escalation reply action: ${action}`);
    }
    if (action === 'continue' && content === undefined) {
      return;
    }
    if (typeof content === 'string') {
      return;
    }
    if (isPlainObject(content)) {
      return;
    }
    throw makeError(
      'transport_protocol_violation',
      `content must be a string or object for escalation action ${action}`,
    );
  }

  function buildEscalationReplyPayload(
    escalationId: string,
    waitToken: string,
    action: EscalationReplyAction,
    content: EscalationReplyContent | undefined,
    meta: Record<string, unknown> | undefined,
  ): Record<string, unknown> {
    validateEscalationReply(escalationId, waitToken, action, content);
    const payload: Record<string, unknown> = {
      escalation_id: escalationId.trim(),
      action,
      wait_token: waitToken.trim(),
    };
    if (content !== undefined) {
      payload['content'] = content;
    }
    if (meta !== undefined) {
      payload['meta'] = meta;
    }
    return payload;
  }

  function handleSystemError(payload: Record<string, unknown>) {
    const code = typeof payload['code'] === 'string' ? payload['code'] : 'session_terminal';
    const message = typeof payload['message'] === 'string' ? payload['message'] : 'Runtime error';
    const entry = lookupError(code);
    const fatal = entry?.fatal ?? false;
    const err = makeError(code, message);

    if (fatal) {
      callbacks.onFatalError(err);
    }
    // Pass through to user as a message regardless
    // (the envelope was already dispatched above via onMessage)
  }

  return {
    setTenantId(tenantId: string | null | undefined) {
      const normalized = typeof tenantId === 'string' ? tenantId.trim() : '';
      _tenantId = normalized || null;
    },

    setTransport(transport: Transport, sendTimeoutMs: number) {
      _transport = transport;
      _sendTimeoutMs = sendTimeoutMs;
    },

    reset,

    get sessionId() { return _sessionId; },
    get sessionState() { return _sessionState; },
    get lastSeq() { return _lastSeq; },

    sendInit(bootstrap: RuntimeBootstrap): Promise<void> {
      _sessionId = null;
      _opened = false;
      _sessionState = 'INITIALIZING';
      // system::init has no session_id — intentionally omit it
      const envelope: Record<string, unknown> = {
        type: 'system::init',
        schema: SCHEMA_VERSION,
        payload: bootstrap,
        meta: { client_msg_id: makeClientMsgId('cli_init') },
        ts: new Date().toISOString(),
      };
      if (_tenantId) envelope['tenant_id'] = _tenantId;
      return send(envelope);
    },

    sendResync(): Promise<void> {
      return send(buildEnvelope('system::resync', { last_seq: _lastSeq }));
    },

    sendStop(): Promise<void> {
      return send(buildEnvelope('sandbox::stop', {}));
    },

    sendChatMessage(content: unknown, attachments: unknown[] | undefined, meta?: Record<string, unknown>): Promise<void> {
      return send(buildEnvelope('chat::message', buildMessagePayload(content, attachments, meta)));
    },

    sendEscalationReply(
      escalationId: string,
      waitToken: string,
      action: EscalationReplyAction,
      content: EscalationReplyContent | undefined,
      meta: Record<string, unknown> | undefined,
    ): Promise<void> {
      return send(buildEnvelope(
        'escalation::reply',
        buildEscalationReplyPayload(escalationId, waitToken, action, content, meta),
      ));
    },

    sendSystemTrigger(content: unknown, attachments: unknown[] | undefined): Promise<void> {
      const payload: Record<string, unknown> = { content };
      if (attachments && attachments.length > 0) {
        payload['meta'] = { attachments };
      }
      return send(buildEnvelope('system::trigger', payload));
    },

    sendTrigger(payload: Record<string, unknown>): Promise<void> {
      return send({
        type: 'system::trigger',
        schema: SCHEMA_VERSION,
        payload: payload || {},
        meta: { client_msg_id: makeClientMsgId('cli_trigger') },
        ts: new Date().toISOString(),
      });
    },

    handleIncoming(data: string) {
      let msg: CortexMessage;
      try {
        msg = JSON.parse(data) as CortexMessage;
      } catch {
        return; // malformed frame — ignore
      }

      if (!_opened) {
        if (msg.type === 'system::opened') {
          if (!msg.session_id) {
            callbacks.onFatalError(
              makeError('transport_protocol_violation', 'system::opened requires session_id'),
            );
            return;
          }
          _sessionId = msg.session_id;
          _opened = true;
          if (!TERMINAL_STATES.has(_sessionState)) {
            _sessionState = 'ACTIVE';
          }
        } else if (msg.type !== 'system::error') {
          callbacks.onFatalError(
            makeError(
              'transport_protocol_violation',
              `Received ${msg.type} before system::opened`,
            ),
          );
          return;
        }
      }

      // Track sequence
      if (typeof msg.seq === 'number' && msg.seq > _lastSeq) {
        _lastSeq = msg.seq;
      }

      updateSessionStateFromMessage(msg);

      // Handle state-changing message types
      if (msg.type === 'system::error') {
        handleSystemError(msg.payload);
      }

      callbacks.onMessage(msg);
    },
  };
}
